File: 242.valid-anagram.js
/*
 * @lc app=leetcode id=242 lang=javascript
 *
 * [242] Valid Anagram
 */

// @lc code=start
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isAnagram = function(s, t) {
    if(s.length !== t.length)
        return false
    let map = {}
    for(let i=0,l=s.length; i<l; i++) {
        let s_c = s[i]
        let t_c = t[i]
        if(!map[s_c])  {
            map[s_c] = 1
        } else {
            map[s_c] ++
        }
        if(!map[t_c]) {
            map[t_c] = -1
        } else {
            map[t_c] --
        }
    }
    for(k in map){
        if(map[k] !== 0){
            return false
        }
    }
    return true
}
isAnagram("anagram", "nagaram")
// @lc code=end

File: 242.valid-anagram.cpp
#include <iostream>
#include <string>
using namespace std;
/*
 * @lc app=leetcode id=242 lang=cpp
 *
 * [242] Valid Anagram
 */

// @lc code=start
class Solution {
public:
    bool isAnagram(string s, string t) {
        if(s.length()!=t.length())
            return false;
        
        int mapS[26]={0}, mapT[26]={0};
        
        for(int i=0;i<s.length();i++){
            cout << "si:" << s[i] << endl;
            cout << "si-a:" << s[i]-'a' << endl;
            mapS[s[i]-'a']++;
            mapT[t[i]-'a']++;
        }

        for(int i=0;i<26;i++){
            if(mapS[i]!=mapT[i])
                return false;
        }
        return true;
    }
};
// @lc code=end

int main()
{
    Solution s;
    bool ans;
    ans = s.isAnagram("aabb", "abcb");
    cout << ans << endl;
    system("pause");
}